File: audit-curriculum.py
#!/usr/bin/env python3
import json
import re

# Load curriculum
with open('src/data/curriculum.json', 'r') as f:
    curriculum = json.load(f)

# Grade level expectations
grade_expectations = {
    0: {'max_number': 20, 'topics': ['Counting', 'Shapes', 'Patterns'], 'no_operations': True},
    1: {'max_number': 100, 'topics': ['Addition', 'Subtraction', 'Counting', 'Time', 'Money'], 'no_mult_div': True},
    2: {'max_number': 1000, 'topics': ['Addition', 'Subtraction', 'Place Value', 'Measurement'], 'no_mult_div': True},
    3: {'max_number': 10000, 'topics': ['Multiplication', 'Division', 'Fractions', 'Area']},
    4: {'max_number': 1000000, 'topics': ['Multiplication', 'Division', 'Fractions', 'Decimals', 'Geometry']},
    5: {'max_number': 1000000, 'topics': ['Decimals', 'Fractions', 'Volume', 'Coordinate Plane']},
    6: {'max_number': float('inf'), 'topics': ['Ratios', 'Percentages', 'Negative Numbers', 'Expressions']}
}

issues = []
warnings = []

print("="*80)
print("CURRICULUM AGE-APPROPRIATENESS AUDIT")
print("="*80)
print(f"\nAnalyzing {len(curriculum)} curriculum sets...\n")

for set_data in curriculum:
    grade = set_data['grade_level']
    expectations = grade_expectations.get(grade, {})
    
    for q_idx, question in enumerate(set_data['questions'], 1):
        q_text = question['question']
        
        # Extract numbers
        numbers = [int(n) for n in re.findall(r'\d+', q_text)]
        max_num = max(numbers) if numbers else 0
        
        # Check number size
        if max_num > expectations.get('max_number', float('inf')):
            issues.append({
                'set': set_data['title'],
                'grade': f"Grade {grade}" if grade > 0 else "Kindergarten",
                'question_num': q_idx,
                'question': q_text[:80] + "..." if len(q_text) > 80 else q_text,
                'issue': f"Number {max_num} too large (max: {expectations['max_number']})",
                'severity': 'HIGH'
            })
        
        # Check for inappropriate operations
        if expectations.get('no_operations') and re.search(r'[\+\-\*÷×]', q_text):
            issues.append({
                'set': set_data['title'],
                'grade': f"Grade {grade}" if grade > 0 else "Kindergarten",
                'question_num': q_idx,
                'question': q_text[:80] + "..." if len(q_text) > 80 else q_text,
                'issue': 'Arithmetic operations not appropriate for Kindergarten',
                'severity': 'HIGH'
            })
        
        if expectations.get('no_mult_div') and re.search(r'[\*÷×]|multipl|divid|times', q_text.lower()):
            issues.append({
                'set': set_data['title'],
                'grade': f"Grade {grade}" if grade > 0 else "Kindergarten",
                'question_num': q_idx,
                'question': q_text[:80] + "..." if len(q_text) > 80 else q_text,
                'issue': f'Multiplication/division not appropriate for Grade {grade}',
                'severity': 'HIGH'
            })
        
        # Check for fractions/decimals in early grades
        if grade <= 2 and re.search(r'fraction|decimal|½|¼|\.5|percent', q_text.lower()):
            issues.append({
                'set': set_data['title'],
                'grade': f"Grade {grade}" if grade > 0 else "Kindergarten",
                'question_num': q_idx,
                'question': q_text[:80] + "..." if len(q_text) > 80 else q_text,
                'issue': f'Fractions/decimals not appropriate for Grade {grade}',
                'severity': 'HIGH'
            })
        
        # Check question length
        word_count = len(q_text.split())
        max_words = 10 if grade == 0 else grade * 8
        if word_count > max_words:
            warnings.append({
                'set': set_data['title'],
                'grade': f"Grade {grade}" if grade > 0 else "Kindergarten",
                'question_num': q_idx,
                'issue': f'Question too long ({word_count} words, max recommended: {max_words})'
            })

# Print results
print(f"✅ Total Sets Analyzed: {len(curriculum)}")
print(f"🚨 High Priority Issues: {len(issues)}")
print(f"⚠️  Warnings: {len(warnings)}\n")

if issues:
    print("\n" + "="*80)
    print("🚨 HIGH PRIORITY ISSUES (MUST FIX)")
    print("="*80)
    for i, issue in enumerate(issues[:20], 1):  # Show first 20
        print(f"\n{i}. {issue['set']} ({issue['grade']})")
        print(f"   Question #{issue['question_num']}: {issue['question']}")
        print(f"   ❌ {issue['issue']}")
    
    if len(issues) > 20:
        print(f"\n... and {len(issues) - 20} more issues")

if warnings:
    print("\n" + "="*80)
    print("⚠️  WARNINGS (Review Recommended)")
    print("="*80)
    for i, warning in enumerate(warnings[:10], 1):  # Show first 10
        print(f"\n{i}. {warning['set']} ({warning['grade']})")
        print(f"   Question #{warning['question_num']}")
        print(f"   ⚠️  {warning['issue']}")
    
    if len(warnings) > 10:
        print(f"\n... and {len(warnings) - 10} more warnings")

# Grade distribution
grade_dist = {}
for set_data in curriculum:
    grade = f"Grade {set_data['grade_level']}" if set_data['grade_level'] > 0 else "Kindergarten"
    grade_dist[grade] = grade_dist.get(grade, 0) + 1

print("\n" + "="*80)
print("📊 GRADE DISTRIBUTION")
print("="*80)
for grade in sorted(grade_dist.keys()):
    print(f"{grade}: {grade_dist[grade]} sets")

# Save report
report = {
    'summary': {
        'total_sets': len(curriculum),
        'total_issues': len(issues),
        'total_warnings': len(warnings)
    },
    'issues': issues,
    'warnings': warnings,
    'grade_distribution': grade_dist
}

with open('audit-report.json', 'w') as f:
    json.dump(report, f, indent=2)

print("\n" + "="*80)
print("💾 Full report saved to: audit-report.json")
print("="*80)

File: audit-curriculum.js
import curriculumData from '../data/curriculum.json';

/**
 * Audit curriculum for age-appropriateness
 * Checks if questions match expected difficulty for grade level
 */

const gradeExpectations = {
    0: { // Kindergarten
        topics: ['Counting', 'Shapes', 'Patterns', 'Comparing'],
        maxNumber: 20,
        operations: ['counting', 'recognition'],
        vocabulary: 'simple, concrete, visual'
    },
    1: { // Grade 1
        topics: ['Addition', 'Subtraction', 'Counting', 'Time', 'Money', 'Measurement'],
        maxNumber: 100,
        operations: ['addition within 20', 'subtraction within 20'],
        vocabulary: 'basic math terms'
    },
    2: { // Grade 2
        topics: ['Addition', 'Subtraction', 'Place Value', 'Measurement', 'Time', 'Money', 'Data'],
        maxNumber: 1000,
        operations: ['addition within 100', 'subtraction within 100', 'skip counting'],
        vocabulary: 'expanded math vocabulary'
    },
    3: { // Grade 3
        topics: ['Multiplication', 'Division', 'Fractions', 'Area', 'Perimeter'],
        maxNumber: 10000,
        operations: ['multiplication tables', 'division basics', 'simple fractions'],
        vocabulary: 'abstract concepts introduced'
    },
    4: { // Grade 4
        topics: ['Multi-digit Multiplication', 'Division', 'Fractions', 'Decimals', 'Geometry'],
        maxNumber: 1000000,
        operations: ['multi-digit operations', 'equivalent fractions', 'decimal basics'],
        vocabulary: 'more complex terminology'
    },
    5: { // Grade 5
        topics: ['Decimals', 'Fractions', 'Volume', 'Coordinate Plane', 'Division'],
        maxNumber: 1000000,
        operations: ['fraction operations', 'decimal operations', 'multi-step problems'],
        vocabulary: 'advanced concepts'
    },
    6: { // Grade 6
        topics: ['Ratios', 'Percentages', 'Negative Numbers', 'Expressions', 'Equations'],
        maxNumber: Infinity,
        operations: ['algebraic thinking', 'ratio reasoning', 'statistical thinking'],
        vocabulary: 'pre-algebra concepts'
    }
};

const issues = [];
const warnings = [];
const recommendations = [];

// Analyze each curriculum set
curriculumData.forEach((set, setIndex) => {
    const grade = set.grade_level;
    const expectations = gradeExpectations[grade];

    if (!expectations) {
        issues.push({
            set: set.id,
            title: set.title,
            issue: `Invalid grade level: ${grade}`,
            severity: 'HIGH'
        });
        return;
    }

    // Check if topic is appropriate for grade
    const topicAppropriate = expectations.topics.some(t =>
        set.topic.toLowerCase().includes(t.toLowerCase()) ||
        t.toLowerCase().includes(set.topic.toLowerCase())
    );

    if (!topicAppropriate && set.topic !== 'Word Problems') {
        warnings.push({
            set: set.id,
            title: set.title,
            grade: grade,
            issue: `Topic "${set.topic}" may not be standard for grade ${grade === 0 ? 'K' : grade}`,
            expected: expectations.topics.join(', '),
            severity: 'MEDIUM'
        });
    }

    // Analyze questions
    set.questions.forEach((q, qIndex) => {
        const questionText = q.question.toLowerCase();

        // Extract numbers from question
        const numbers = questionText.match(/\d+/g)?.map(Number) || [];
        const maxNum = Math.max(...numbers, 0);

        // Check if numbers are too large for grade level
        if (maxNum > expectations.maxNumber) {
            issues.push({
                set: set.id,
                title: set.title,
                grade: grade,
                questionIndex: qIndex + 1,
                question: q.question,
                issue: `Number ${maxNum} exceeds grade ${grade === 0 ? 'K' : grade} expectation (max: ${expectations.maxNumber})`,
                severity: 'HIGH'
            });
        }

        // Check for grade-inappropriate operations
        if (grade === 0) {
            // Kindergarten shouldn't have arithmetic operations
            if (questionText.match(/\+|-|\*|÷|×/)) {
                issues.push({
                    set: set.id,
                    title: set.title,
                    grade: grade,
                    questionIndex: qIndex + 1,
                    question: q.question,
                    issue: 'Kindergarten should not have arithmetic symbols',
                    severity: 'HIGH'
                });
            }
        }

        if (grade === 1) {
            // Grade 1 shouldn't have multiplication/division
            if (questionText.match(/\*|÷|×|multiply|divide|times/)) {
                issues.push({
                    set: set.id,
                    title: set.title,
                    grade: grade,
                    questionIndex: qIndex + 1,
                    question: q.question,
                    issue: 'Grade 1 should not have multiplication/division',
                    severity: 'HIGH'
                });
            }
        }

        if (grade <= 2) {
            // Early grades shouldn't have fractions/decimals
            if (questionText.match(/fraction|decimal|\.5|½|¼|\/|percent/)) {
                issues.push({
                    set: set.id,
                    title: set.title,
                    grade: grade,
                    questionIndex: qIndex + 1,
                    question: q.question,
                    issue: `Grade ${grade} should not have fractions/decimals`,
                    severity: 'HIGH'
                });
            }
        }

        // Check vocabulary complexity
        const complexWords = [
            'coefficient', 'polynomial', 'quadratic', 'exponent', 'logarithm',
            'derivative', 'integral', 'theorem', 'proof', 'congruent'
        ];

        complexWords.forEach(word => {
            if (questionText.includes(word) && grade < 6) {
                issues.push({
                    set: set.id,
                    title: set.title,
                    grade: grade,
                    questionIndex: qIndex + 1,
                    question: q.question,
                    issue: `Word "${word}" is too advanced for grade ${grade === 0 ? 'K' : grade}`,
                    severity: 'MEDIUM'
                });
            }
        });

        // Check question length (age-appropriate reading level)
        const wordCount = q.question.split(' ').length;
        const maxWords = grade === 0 ? 10 : grade * 8;

        if (wordCount > maxWords) {
            warnings.push({
                set: set.id,
                title: set.title,
                grade: grade,
                questionIndex: qIndex + 1,
                question: q.question,
                issue: `Question too long (${wordCount} words) for grade ${grade === 0 ? 'K' : grade} (max recommended: ${maxWords})`,
                severity: 'LOW'
            });
        }
    });
});

// Generate report
console.log('='.repeat(80));
console.log('CURRICULUM AGE-APPROPRIATENESS AUDIT REPORT');
console.log('='.repeat(80));
console.log(`\nTotal Curriculum Sets Analyzed: ${curriculumData.length}`);
console.log(`Total Issues Found: ${issues.length}`);
console.log(`Total Warnings: ${warnings.length}\n`);

if (issues.length > 0) {
    console.log('\n🚨 HIGH PRIORITY ISSUES (Must Fix):');
    console.log('-'.repeat(80));
    issues.filter(i => i.severity === 'HIGH').forEach((issue, idx) => {
        console.log(`\n${idx + 1}. ${issue.title} (${issue.set})`);
        console.log(`   Grade: ${issue.grade === 0 ? 'K' : issue.grade}`);
        if (issue.questionIndex) console.log(`   Question #${issue.questionIndex}: "${issue.question}"`);
        console.log(`   ❌ ${issue.issue}`);
    });
}

if (warnings.length > 0) {
    console.log('\n\n⚠️  WARNINGS (Review Recommended):');
    console.log('-'.repeat(80));
    warnings.slice(0, 10).forEach((warning, idx) => {
        console.log(`\n${idx + 1}. ${warning.title} (${warning.set})`);
        console.log(`   Grade: ${warning.grade === 0 ? 'K' : warning.grade}`);
        if (warning.questionIndex) console.log(`   Question #${warning.questionIndex}`);
        console.log(`   ⚠️  ${warning.issue}`);
    });
    if (warnings.length > 10) {
        console.log(`\n   ... and ${warnings.length - 10} more warnings`);
    }
}

// Export results
const report = {
    summary: {
        totalSets: curriculumData.length,
        totalIssues: issues.length,
        totalWarnings: warnings.length,
        highPriorityIssues: issues.filter(i => i.severity === 'HIGH').length,
        mediumPriorityIssues: issues.filter(i => i.severity === 'MEDIUM').length
    },
    issues,
    warnings,
    gradeDistribution: curriculumData.reduce((acc, set) => {
        const grade = set.grade_level === 0 ? 'K' : `Grade ${set.grade_level}`;
        acc[grade] = (acc[grade] || 0) + 1;
        return acc;
    }, {})
};

console.log('\n\n📊 GRADE DISTRIBUTION:');
console.log('-'.repeat(80));
Object.entries(report.gradeDistribution).sort().forEach(([grade, count]) => {
    console.log(`${grade}: ${count} sets`);
});

console.log('\n\n💾 Full report saved to: audit-report.json');
console.log('='.repeat(80));

// Save to file
import fs from 'fs';
fs.writeFileSync('audit-report.json', JSON.stringify(report, null, 2));
